src/librustc_typeck/collect.rs

```diff
@@ -2157,24 +2157,39 @@ fn explicit_predicates_of<'tcx>(
     if let Some((self_trait_ref, trait_items)) = is_trait {
         predicates.extend(trait_items.iter().flat_map(|trait_item_ref| {
             let trait_item = tcx.hir().trait_item(trait_item_ref.id);
-            let bounds = match trait_item.node {
-                hir::TraitItemKind::Type(ref bounds, _) => bounds,
+            let (bounds, default) = match trait_item.node {
+                hir::TraitItemKind::Type(ref bounds, ref default) => (bounds, default),
                 _ => return Vec::new().into_iter()
             };
 
-            let assoc_ty =
-                tcx.mk_projection(tcx.hir().local_def_id_from_hir_id(trait_item.hir_id),
-                    self_trait_ref.substs);
+            let did = tcx.hir().local_def_id_from_hir_id(trait_item.hir_id);
+            let assoc_ty = tcx.mk_projection(did, self_trait_ref.substs);
 
-            let bounds = AstConv::compute_bounds(
+            let mut predicates = AstConv::compute_bounds(
                 &ItemCtxt::new(tcx, def_id),
                 assoc_ty,
                 bounds,
                 SizedByDefault::Yes,
                 trait_item.span,
-            );
+            ).predicates(tcx, assoc_ty);
+
+            if let Some(default_ty) = default {
+                let default_ty = AstConv::ast_ty_to_ty(&ItemCtxt::new(tcx, def_id), default_ty);
+                let preds = AstConv::compute_bounds(
```

Comment:
Can you add a test to also make sure that bounds are respected when not written directly on the associated type itself?

```rust
trait Foo<T>
where
    // Same bound as `type Bar: Clone` but written indirectly:
    <Self as Foo<T>>::Bar: Clone,
{
    type Bar = Vec<T>;
}
```

Reply:
Another case to consider is when you have multiple defaults:

```rust
trait Foo<T>
where
    <Self as Foo<T>>::Bar: Clone,
{
    type Bar = Vec<Self::Baz>;
    type Baz = T;
}
```

Reply:
Yeah this doesn't work at all right now, this check needs to be implemented differently.

This gets particularly tricky when you have a where-clause like `where Self::Ty: Con<Self>` on the trait. You can't really know if this holds for the defaulted type because `Self` isn't yet known.

Reply:
In general, when faced with a situation "you can't really know yet" this should be treated as if it wasn't so. In other words, we want constructive proof that something holds. So in a situation with `where Self::Ty: Con<Self>`, because `Self` isn't yet known, the answer is to treat it as a type variable (which it is...) and then error.

I think the check should be something along the lines of substituting the associated types for the defaults and then try to prove the normalized `where` clauses. It seems to me that rejecting `where Self::Ty: Con<Self>,` should just "fall out".

Reply:
Just pushed a better implementation of the check and a comprehensive test

Reply:
Looks great!

Can you also add my last snippet to that so we check for scenarios involving more than one associated type?

Reply:
Done. Right now the behavior is to do a shallow replacement of associated types with their defaults, so `<Self as Foo<T>>::Bar: Clone` is turned into `Vec<Self::Baz>`, not `Vec<T>`. I guess this needs to be fixed.

Reply:
Ah, I think the shallow substitution is actually beneficial here, since it forces the trait to work with any combination of defaults and non-defaults: If we did the deep substitution and had a `T: Clone` bound, we'd prove that `Vec<T>: Clone` and the trait would be accepted. If a user then implements it and overrides *only* `Baz`, suddenly `Vec<Self::Baz>: Clone` might no longer hold, and the user would be forced to also override `Bar`. With the shallow substitution, the trait would have to add a `Self::Baz: Clone` bound instead/in addition to the `T: Clone` bound, which means that an implementation could replace either assoc. type without making others not work anymore.

Reply:
Interesting; it seems to me a shallow substitution is strictly more conservative than the deeper one, right? So we can relax it later if we want.

I think it would be good to demonstrate the difference in a special test file and then also leave a description of the distinction (with examples) on the tracking issue so that these interesting aspects aren't lost in comments here.